File: Tween/Growl.vb
Imports System.Reflection
Imports System.IO
Imports System.ComponentModel
Imports System.Drawing.Imaging

Public Class GrowlHelper

    Private _connector As Assembly = Nothing
    Private _core As Assembly = Nothing

    Private _growlNTreply As Object
    Private _growlNTdm As Object
    Private _growlNTnew As Object
    Private _growlNTusevent As Object
    Private _growlApp As Object

    Private _targetConnector As Object
    Private _targetCore As Object

    Private _appName As String = ""
    Dim _initialized As Boolean = False

    Public Class NotifyCallbackEventArgs
        Inherits EventArgs
        Public Property StatusId As Long
        Public Property NotifyType As NotifyType
        Public Sub New(ByVal notifyType As NotifyType, ByVal statusId As String)
            If statusId.Length > 1 Then
                Me.StatusId = CLng(statusId)
                Me.NotifyType = notifyType
            End If
        End Sub
    End Class

    Public Event NotifyClicked(ByVal sender As Object, ByVal e As NotifyCallbackEventArgs)

    Public ReadOnly Property AppName As String
        Get
            Return _appName
        End Get
    End Property

    Public Enum NotifyType
        Reply = 0
        DirectMessage = 1
        Notify = 2
        UserStreamEvent = 3
    End Enum

    Public Sub New(ByVal appName As String)
        _appName = appName
    End Sub

    Public ReadOnly Property IsAvailable As Boolean
        Get
            If _connector Is Nothing OrElse _core Is Nothing OrElse Not _initialized Then
                Return False
            Else
                Return True
            End If
        End Get
    End Property

    Private Overloads Function IconToByteArray(ByVal filename As String) As Byte()
        Return IconToByteArray(New Icon(filename))
    End Function

    Private Overloads Function IconToByteArray(ByVal icondata As Icon) As Byte()
        Using ms As New MemoryStream
            Dim ic As Icon = New Icon(icondata, 48, 48)
            ic.ToBitmap.Save(ms, ImageFormat.Png)
            Return ms.ToArray()
        End Using
    End Function

    Public Shared ReadOnly Property IsDllExists As Boolean
        Get
            Dim dir As String = Application.StartupPath
            Dim connectorPath As String = Path.Combine(dir, "Growl.Connector.dll")
            Dim corePath As String = Path.Combine(dir, "Growl.CoreLibrary.dll")
            If File.Exists(connectorPath) AndAlso File.Exists(corePath) Then
                Return True
            Else
                Return False
            End If
        End Get
    End Property

    Public Function RegisterGrowl() As Boolean

        _initialized = False
        Dim dir As String = Application.StartupPath
        Dim connectorPath As String = Path.Combine(dir, "Growl.Connector.dll")
        Dim corePath As String = Path.Combine(dir, "Growl.CoreLibrary.dll")

        Try
            If Not IsDllExists Then Return False
            _connector = Assembly.LoadFile(connectorPath)
            _core = Assembly.LoadFile(corePath)
        Catch ex As Exception
            Return False
        End Try

        Try
            _targetConnector = _connector.CreateInstance("Growl.Connector.GrowlConnector")
            _targetCore = _core.CreateInstance("Growl.CoreLibrary")
            Dim _t As Type = _connector.GetType("Growl.Connector.NotificationType")

            _growlNTreply = _t.InvokeMember(Nothing,
                BindingFlags.CreateInstance, Nothing, Nothing, New Object() {"REPLY", "Reply"})

            _growlNTdm = _t.InvokeMember(Nothing,
                BindingFlags.CreateInstance, Nothing, Nothing, New Object() {"DIRECT_MESSAGE", "DirectMessage"})

            _growlNTnew = _t.InvokeMember(Nothing,
                BindingFlags.CreateInstance, Nothing, Nothing, New Object() {"NOTIFY", "新着通知"})

            _growlNTusevent = _t.InvokeMember(Nothing,
                BindingFlags.CreateInstance, Nothing, Nothing, New Object() {"USERSTREAM_EVENT", "UserStream Event"})

            Dim encryptType As Object =
                    _connector.GetType("Growl.Connector.Cryptography+SymmetricAlgorithmType").InvokeMember(
                        "PlainText", BindingFlags.GetField, Nothing, Nothing, Nothing)
            _targetConnector.GetType.InvokeMember("EncryptionAlgorithm", BindingFlags.SetProperty, Nothing, _targetConnector, New Object() {encryptType})

            _growlApp = _connector.CreateInstance(
                "Growl.Connector.Application", False, BindingFlags.Default, Nothing, New Object() {_appName}, Nothing, Nothing)


            If File.Exists(Path.Combine(Application.StartupPath, "Icons\Tween.png")) Then
                ' Icons\Tween.pngを使用
                Dim ci As ConstructorInfo = _core.GetType(
                    "Growl.CoreLibrary.Resource").GetConstructor(
                    BindingFlags.NonPublic Or BindingFlags.Instance,
                    Nothing, New Type() {GetType(System.String)}, Nothing)

                Dim data As Object = ci.Invoke(New Object() {Path.Combine(Application.StartupPath, "Icons\Tween.png")})
                Dim pi As PropertyInfo = _growlApp.GetType.GetProperty("Icon")
                pi.SetValue(_growlApp, data, Nothing)

            ElseIf File.Exists(Path.Combine(Application.StartupPath, "Icons\MIcon.ico")) Then
                ' アイコンセットにMIcon.icoが存在する場合それを使用
                Dim cibd As ConstructorInfo = _core.GetType(
                     "Growl.CoreLibrary.BinaryData").GetConstructor(
                     BindingFlags.Public Or BindingFlags.Instance,
                     Nothing, New Type() {GetType(Byte())}, Nothing)
                Dim tc As New TypeConverter
                Dim bdata As Object = cibd.Invoke(
                    New Object() {IconToByteArray(Path.Combine(Application.StartupPath, "Icons\MIcon.ico"))})

                Dim ciRes As ConstructorInfo = _core.GetType(
                    "Growl.CoreLibrary.Resource").GetConstructor(
                    BindingFlags.NonPublic Or BindingFlags.Instance,
                    Nothing, New Type() {bdata.GetType()}, Nothing)

                Dim data As Object = ciRes.Invoke(New Object() {bdata})
                Dim pi As PropertyInfo = _growlApp.GetType.GetProperty("Icon")
                pi.SetValue(_growlApp, data, Nothing)
            Else
                '内蔵アイコンリソースを使用
                Dim cibd As ConstructorInfo = _core.GetType(
                     "Growl.CoreLibrary.BinaryData").GetConstructor(
                     BindingFlags.Public Or BindingFlags.Instance,
                     Nothing, New Type() {GetType(Byte())}, Nothing)
                Dim tc As New TypeConverter
                Dim bdata As Object = cibd.Invoke(
                    New Object() {IconToByteArray(My.Resources.MIcon)})

                Dim ciRes As ConstructorInfo = _core.GetType(
                    "Growl.CoreLibrary.Resource").GetConstructor(
                    BindingFlags.NonPublic Or BindingFlags.Instance,
                    Nothing, New Type() {bdata.GetType()}, Nothing)

                Dim data As Object = ciRes.Invoke(New Object() {bdata})
                Dim pi As PropertyInfo = _growlApp.GetType.GetProperty("Icon")
                pi.SetValue(_growlApp, data, Nothing)
            End If

            Dim mi As MethodInfo = _targetConnector.GetType.GetMethod("Register", New Type() {_growlApp.GetType, _connector.GetType("Growl.Connector.NotificationType[]")})

            _t = _connector.GetType("Growl.Connector.NotificationType")

            Dim arglist As New ArrayList
            arglist.Add(_growlNTreply)
            arglist.Add(_growlNTdm)
            arglist.Add(_growlNTnew)
            arglist.Add(_growlNTusevent)

            mi.Invoke(_targetConnector, New Object() {_growlApp, arglist.ToArray(_t)})

            ' コールバックメソッドの登録
            Dim tGrowlConnector As Type = _connector.GetType("Growl.Connector.GrowlConnector")
            Dim evNotificationCallback As EventInfo = tGrowlConnector.GetEvent("NotificationCallback")
            Dim tDelegate As Type = evNotificationCallback.EventHandlerType
            Dim miHandler As MethodInfo = GetType(GrowlHelper).GetMethod("GrowlCallbackHandler", BindingFlags.NonPublic Or BindingFlags.Instance)
            Dim d As [Delegate] = [Delegate].CreateDelegate(tDelegate, Me, miHandler)
            Dim miAddHandler As MethodInfo = evNotificationCallback.GetAddMethod()
            Dim addHandlerArgs() As Object = {d}
            miAddHandler.Invoke(_targetConnector, addHandlerArgs)

            _initialized = True

        Catch ex As Exception
            _initialized = False
            Return False
        End Try

        Return True
    End Function

    Public Sub Notify(ByVal notificationType As NotifyType, ByVal id As String, ByVal title As String, ByVal text As String, Optional ByVal icon As Image = Nothing, Optional ByVal url As String = "")
        If Not _initialized Then Return
        Dim notificationName As String = ""
        Select Case notificationType
            Case NotifyType.Reply
                notificationName = "REPLY"
            Case NotifyType.DirectMessage
                notificationName = "DIRECT_MESSAGE"
            Case NotifyType.Notify
                notificationName = "NOTIFY"
            Case NotifyType.UserStreamEvent
                notificationName = "USERSTREAM_EVENT"
        End Select
        Dim n As Object = Nothing
        If icon IsNot Nothing OrElse Not String.IsNullOrEmpty(url) Then
            Dim gCore As Type = _core.GetType("Growl.CoreLibrary.Resource")
            Dim res As Object = Nothing
            If icon IsNot Nothing Then
                res = gCore.InvokeMember("op_Implicit",
                                                   BindingFlags.Public Or BindingFlags.Static Or BindingFlags.InvokeMethod,
                                                   Nothing,
                                                   Nothing,
                                                   New Object() {icon})
            Else
                res = gCore.InvokeMember("op_Implicit",
                                                   BindingFlags.Public Or BindingFlags.Static Or BindingFlags.InvokeMethod,
                                                   Nothing,
                                                   Nothing,
                                                   New Object() {url})
            End If
            Dim priority As Object =
                    _connector.GetType("Growl.Connector.Priority").InvokeMember(
                        "Normal", BindingFlags.GetField, Nothing, Nothing, Nothing)
            n = _connector.GetType("Growl.Connector.Notification").InvokeMember(
                    "Notification",
                    BindingFlags.CreateInstance,
                    Nothing,
                    _connector,
                    New Object() {_appName,
                                  notificationName,
                                  id,
                                  title,
                                  text,
                                  res,
                                  False,
                                  priority,
                                  "aaa"})
        Else
            n = _connector.GetType("Growl.Connector.Notification").InvokeMember(
                    "Notification",
                    BindingFlags.CreateInstance,
                    Nothing,
                    _connector,
                    New Object() {_appName,
                                    notificationName,
                                    id,
                                    title,
                                    text})
        End If
        '_targetConnector.GetType.InvokeMember("Notify", BindingFlags.InvokeMethod, Nothing, _targetConnector, New Object() {n})
        Dim cc As Object = _connector.GetType("Growl.Connector.CallbackContext").InvokeMember(
            Nothing, BindingFlags.CreateInstance, Nothing, _connector,
            New Object() {"some fake information", notificationName})
        _targetConnector.GetType.InvokeMember("Notify", BindingFlags.InvokeMethod, Nothing, _targetConnector, New Object() {n, cc})
    End Sub

    Private Sub GrowlCallbackHandler(ByVal response As Object, ByVal callbackData As Object, ByVal state As Object)
        Try
            ' 定数取得
            Dim vCLICK As Object =
            _core.GetType("Growl.CoreLibrary.CallbackResult").GetField(
                        "CLICK",
                       BindingFlags.Public Or BindingFlags.Static).GetRawConstantValue()
            ' 実際の値
            Dim vResult As Object = callbackData.GetType.GetProperty(
                        "Result",
                        BindingFlags.Public Or BindingFlags.Instance).GetGetMethod.Invoke(callbackData, Nothing)
            vResult = CType(vResult, Integer)
            Dim notifyId As String = CStr(callbackData.GetType.GetProperty("NotificationID").GetGetMethod.Invoke(callbackData, Nothing))
            Dim notifyName As String = CStr(callbackData.GetType.GetProperty("Type").GetGetMethod.Invoke(callbackData, Nothing))
            If vCLICK.Equals(vResult) Then
                Dim nt As NotifyType
                Select Case notifyName
                    Case "REPLY"
                        nt = NotifyType.Reply
                    Case "DIRECT_MESSAGE"
                        nt = NotifyType.DirectMessage
                    Case "NOTIFY"
                        nt = NotifyType.Notify
                    Case "USERSTREAM_EVENT"
                        nt = NotifyType.UserStreamEvent
                End Select
                RaiseEvent NotifyClicked(Me, New NotifyCallbackEventArgs(nt, notifyId))
            End If
        Catch ex As Exception
            Exit Sub
        End Try
    End Sub
End Class

File: OpenTween/Growl.cs
// OpenTween - Client of Twitter
// Copyright (c) 2007-2011 kiri_feather (@kiri_feather) <kiri.feather@gmail.com>
//           (c) 2008-2011 Moz (@syo68k)
//           (c) 2008-2011 takeshik (@takeshik) <http://www.takeshik.org/>
//           (c) 2010-2011 anis774 (@anis774) <http://d.hatena.ne.jp/anis774/>
//           (c) 2010-2011 fantasticswallow (@f_swallow) <http://twitter.com/f_swallow>
//           (c) 2011      kim_upsilon (@kim_upsilon) <https://upsilo.net/~upsilon/>
// All rights reserved.
//
// This file is part of OpenTween.
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General public License as published by the Free
// Software Foundation; either version 3 of the License, or (at your option)
// any later version.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
// or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General public License
// for more details.
//
// You should have received a copy of the GNU General public License along
// with this program. If not, see <http://www.gnu.org/licenses/>, or write to
// the Free Software Foundation, Inc., 51 Franklin Street - Fifth Floor,
// Boston, MA 02110-1301, USA.

#nullable enable

using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Drawing.Imaging;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Windows.Forms;

namespace OpenTween
{
    public class GrowlHelper
    {
        private Assembly? connector = null;
        private Assembly? core = null;

        private object? growlNTreply;
        private object? growlNTdm;
        private object? growlNTnew;
        private object? growlApp;

        private object? targetConnector;
        private bool initialized = false;

        public class NotifyCallbackEventArgs : EventArgs
        {
            public long StatusId { get; set; }
            public NotifyType NotifyType { get; set; }
            public NotifyCallbackEventArgs(NotifyType notifyType, string statusId)
            {
                if (statusId.Length > 1)
                {
                    this.StatusId = long.Parse(statusId);
                    this.NotifyType = notifyType;
                }
            }
        }

        public event EventHandler<NotifyCallbackEventArgs>? NotifyClicked;

        public string AppName { get; }

        public enum NotifyType
        {
            Reply = 0,
            DirectMessage = 1,
            Notify = 2,
        }

        public GrowlHelper(string appName)
            => this.AppName = appName;

        public bool IsAvailable
        {
            get
            {
                if (this.connector == null || this.core == null || !this.initialized)
                    return false;
                else
                    return true;
            }
        }

        private byte[] IconToByteArray(string filename)
        {
            using var ic = new Icon(filename);
            return this.IconToByteArray(ic);
        }

        private byte[] IconToByteArray(Icon icondata)
        {
            using var ms = new MemoryStream();
            using var ic = new Icon(icondata, 48, 48);
            ic.ToBitmap().Save(ms, ImageFormat.Png);
            return ms.ToArray();
        }

        public static bool IsDllExists
        {
            get
            {
                var dir = Application.StartupPath;
                var connectorPath = Path.Combine(dir, "Growl.Connector.dll");
                var corePath = Path.Combine(dir, "Growl.CoreLibrary.dll");
                if (File.Exists(connectorPath) && File.Exists(corePath))
                    return true;
                else
                    return false;
            }
        }

        public bool RegisterGrowl()
        {
            this.initialized = false;
            var dir = Application.StartupPath;
            var connectorPath = Path.Combine(dir, "Growl.Connector.dll");
            var corePath = Path.Combine(dir, "Growl.CoreLibrary.dll");

            try
            {
                if (!IsDllExists) return false;
                this.connector = Assembly.LoadFile(connectorPath);
                this.core = Assembly.LoadFile(corePath);
            }
            catch (Exception)
            {
                return false;
            }

            try
            {
                this.targetConnector = this.connector.CreateInstance("Growl.Connector.GrowlConnector");
                var t = this.connector.GetType("Growl.Connector.NotificationType");

                this.growlNTreply = t.InvokeMember(
                    null,
                    BindingFlags.CreateInstance,
                    null,
                    null,
                    new object[] { "REPLY", "Reply" },
                    CultureInfo.InvariantCulture);

                this.growlNTdm = t.InvokeMember(null,
                    BindingFlags.CreateInstance,
                    null,
                    null,
                    new object[] { "DIRECT_MESSAGE", "DirectMessage" },
                    CultureInfo.InvariantCulture);

                this.growlNTnew = t.InvokeMember(
                    null,
                    BindingFlags.CreateInstance,
                    null,
                    null,
                    new object[] { "NOTIFY", "新着通知" },
                    CultureInfo.InvariantCulture);

                var encryptType =
                        this.connector.GetType("Growl.Connector.Cryptography+SymmetricAlgorithmType").InvokeMember(
                            "PlainText", BindingFlags.GetField, null, null, null, CultureInfo.InvariantCulture);
                this.targetConnector.GetType().InvokeMember("EncryptionAlgorithm", BindingFlags.SetProperty, null, this.targetConnector, new object[] { encryptType }, CultureInfo.InvariantCulture);

                this.growlApp = this.connector.CreateInstance(
                    "Growl.Connector.Application", false, BindingFlags.Default, null, new object[] { this.AppName }, null, null);

                if (File.Exists(Path.Combine(Application.StartupPath, "Icons\\Tween.png")))
                {
                    // Icons\Tween.pngを使用
                    var ci = this.core.GetType("Growl.CoreLibrary.Resource").GetConstructor(
                        BindingFlags.NonPublic | BindingFlags.Instance,
                        null,
                        new Type[] { typeof(string) },
                        null);

                    var data = ci.Invoke(new object[] { Path.Combine(Application.StartupPath, "Icons\\Tween.png") });
                    var pi = this.growlApp.GetType().GetProperty("Icon");
                    pi.SetValue(this.growlApp, data, null);
                }
                else if (File.Exists(Path.Combine(Application.StartupPath, "Icons\\MIcon.ico")))
                {
                    // アイコンセットにMIcon.icoが存在する場合それを使用
                    var cibd = this.core.GetType("Growl.CoreLibrary.BinaryData").GetConstructor(
                        BindingFlags.Public | BindingFlags.Instance,
                        null,
                        new Type[] { typeof(byte[]) },
                        null);
                    var bdata = cibd.Invoke(
                        new object[] { this.IconToByteArray(Path.Combine(Application.StartupPath, "Icons\\MIcon.ico")) });

                    var ciRes = this.core.GetType("Growl.CoreLibrary.Resource").GetConstructor(
                        BindingFlags.NonPublic | BindingFlags.Instance,
                        null,
                        new Type[] { bdata.GetType() },
                        null);

                    var data = ciRes.Invoke(new object[] { bdata });
                    var pi = this.growlApp.GetType().GetProperty("Icon");
                    pi.SetValue(this.growlApp, data, null);
                }
                else
                {
                    // 内蔵アイコンリソースを使用
                    var cibd = this.core.GetType("Growl.CoreLibrary.BinaryData").GetConstructor(
                        BindingFlags.Public | BindingFlags.Instance,
                        null,
                        new Type[] { typeof(byte[]) },
                        null);
                    var bdata = cibd.Invoke(
                        new object[] { this.IconToByteArray(Properties.Resources.MIcon) });

                    var ciRes = this.core.GetType("Growl.CoreLibrary.Resource").GetConstructor(
                        BindingFlags.NonPublic | BindingFlags.Instance,
                        null,
                        new Type[] { bdata.GetType() },
                        null);

                    var data = ciRes.Invoke(new object[] { bdata });
                    var pi = this.growlApp.GetType().GetProperty("Icon");
                    pi.SetValue(this.growlApp, data, null);
                }

                var mi = this.targetConnector.GetType().GetMethod("Register", new Type[] { this.growlApp.GetType(), this.connector.GetType("Growl.Connector.NotificationType[]") });

                t = this.connector.GetType("Growl.Connector.NotificationType");

                var arglist = new ArrayList
                {
                    this.growlNTreply,
                    this.growlNTdm,
                    this.growlNTnew,
                };

                mi.Invoke(this.targetConnector, new object[] { this.growlApp, arglist.ToArray(t) });

                // コールバックメソッドの登録
                var tGrowlConnector = this.connector.GetType("Growl.Connector.GrowlConnector");
                var evNotificationCallback = tGrowlConnector.GetEvent("NotificationCallback");
                var tDelegate = evNotificationCallback.EventHandlerType;
                var miHandler = typeof(GrowlHelper).GetMethod("GrowlCallbackHandler", BindingFlags.NonPublic | BindingFlags.Instance);
                var d = Delegate.CreateDelegate(tDelegate, this, miHandler);
                var miAddHandler = evNotificationCallback.GetAddMethod();
                object[] addHandlerArgs = { d };
                miAddHandler.Invoke(this.targetConnector, addHandlerArgs);

                this.initialized = true;
            }
            catch (Exception)
            {
                this.initialized = false;
                return false;
            }

            return true;
        }

        public void Notify(NotifyType notificationType, string id, string title, string text, Image? icon = null, string url = "")
        {
            if (!this.initialized) return;

            var notificationName = notificationType switch
            {
                NotifyType.Reply => "REPLY",
                NotifyType.DirectMessage => "DIRECT_MESSAGE",
                NotifyType.Notify => "NOTIFY",
                _ => "",
            };

            object? n;
            if (icon != null || !MyCommon.IsNullOrEmpty(url))
            {
                var gCore = this.core!.GetType("Growl.CoreLibrary.Resource");
                object? res;
                if (icon != null)
                {
                    res = gCore.InvokeMember("op_Implicit",
                                             BindingFlags.Public | BindingFlags.Static | BindingFlags.InvokeMethod,
                                             null,
                                             null,
                                             new object[] { icon },
                                             CultureInfo.InvariantCulture);
                }
                else
                {
                    res = gCore.InvokeMember("op_Implicit",
                                             BindingFlags.Public | BindingFlags.Static | BindingFlags.InvokeMethod,
                                             null,
                                             null,
                                             new object[] { url },
                                             CultureInfo.InvariantCulture);
                }
                var priority =
                        this.connector!.GetType("Growl.Connector.Priority").InvokeMember(
                            "Normal", BindingFlags.GetField, null, null, null, CultureInfo.InvariantCulture);
                n = this.connector!.GetType("Growl.Connector.Notification").InvokeMember(
                        "Notification",
                        BindingFlags.CreateInstance,
                        null,
                        this.connector,
                        new object[]
                        {
                            this.AppName,
                            notificationName,
                            id,
                            title,
                            text,
                            res,
                            false,
                            priority,
                            "aaa",
                        },
                        CultureInfo.InvariantCulture);
            }
            else
            {
                n = this.connector!.GetType("Growl.Connector.Notification").InvokeMember(
                        "Notification",
                        BindingFlags.CreateInstance,
                        null,
                        this.connector,
                        new object[]
                        {
                            this.AppName,
                            notificationName,
                            id,
                            title,
                            text,
                        },
                        CultureInfo.InvariantCulture);
            }
            var cc = this.connector.GetType("Growl.Connector.CallbackContext").InvokeMember(
                null,
                BindingFlags.CreateInstance,
                null,
                this.connector,
                new object[] { "some fake information", notificationName },
                CultureInfo.InvariantCulture);
            this.targetConnector!.GetType().InvokeMember("Notify", BindingFlags.InvokeMethod, null, this.targetConnector, new object[] { n, cc }, CultureInfo.InvariantCulture);
        }

        private void GrowlCallbackHandler(object response, object callbackData, object state)
        {
            try
            {
                // 定数取得
                var vCLICK =
                this.core!.GetType("Growl.CoreLibrary.CallbackResult").GetField(
                    "CLICK",
                    BindingFlags.Public | BindingFlags.Static).GetRawConstantValue();
                // 実際の値
                var vResult = callbackData.GetType().GetProperty(
                            "Result",
                            BindingFlags.Public | BindingFlags.Instance).GetGetMethod().Invoke(callbackData, null);
                vResult = (int)vResult;
                var notifyId = (string)callbackData.GetType().GetProperty("NotificationID").GetGetMethod().Invoke(callbackData, null);
                var notifyName = (string)callbackData.GetType().GetProperty("Type").GetGetMethod().Invoke(callbackData, null);
                if (vCLICK.Equals(vResult))
                {
                    NotifyType nt;
                    switch (notifyName)
                    {
                        case "REPLY":
                            nt = NotifyType.Reply;
                            break;
                        case "DIRECT_MESSAGE":
                            nt = NotifyType.DirectMessage;
                            break;
                        case "NOTIFY":
                            nt = NotifyType.Notify;
                            break;
                        default:
                            return;
                    }

                    this.NotifyClicked?.Invoke(this, new NotifyCallbackEventArgs(nt, notifyId));
                }
            }
            catch (Exception)
            {
                return;
            }
        }
    }
}
